xarray/core/dataset.py

```diff
@@ -2748,6 +2748,54 @@ def roll(self, **shifts):
 
         return self._replace_vars_and_dims(variables)
 
+    def sortby(self, variables, ascending=True):
+        """
+        Sorts the dataset, either along specified dimensions,
+        or according to values of 1-D dataarrays that share dimension
+        with calling object.  If multiple sorts along the same dimension is
+        given, numpy's lexsort is performed along that dimension:
+        https://docs.scipy.org/doc/numpy/reference/generated/numpy.lexsort.html
+        and the FIRST key in the sequence is used as the primary sort key,
+        followed by the 2nd key, etc.
+
+
+        Parameters
+        ----------
+        variables: (str, DataArray, or list of either)
+            Name(s) of a 1D variable in coords/data_vars whose values are used
+            to sort the dataset.
+        ascending: boolean, optional.
+            whether to sort by ascending order.
+
+        Returns
+        -------
+        sorted: Dataset
+            A new dataset where all the specified dims are sorted by dim
+            labels.
+        """
+        from .dataarray import DataArray
+
+        if not isinstance(variables, list):
+            vs = [variables]
+        else:
+            vs = variables
+        vs = [v if isinstance(v, DataArray) else self[v] for v in vs]
+
+        dims = {}
+        for d in vs:
+            if len(d.dims) > 1:
+                raise ValueError("Input DataArray has more than 1 dimension.")
+            else:
+                key = d.dims[0]
+                val = d.argsort() if ascending else d.argsort()[::-1]
+                if len(val) != len(self[key]):
+                    raise ValueError("Input DataArray must have same "
+                                     "length as dimension it is sorting.")
+            if key in dims:
+                dims[key] = np.lexsort((val, dims[key]))
```

Comment:
I'm not sure this is calculating the right thing. I think you need to lexsort `d`, not `val` (the result of `argsort`). Otherwise, lexsort does some sort of shuffle of the sort indices:
```
>>> np.lexsort(([2, 1, 0],))
array([2, 1, 0])
```

Maybe [`inverse_permutation`](https://github.com/pydata/xarray/blob/8f6a68e3f821689203bce2bce52b412e9fe70b5c/xarray/core/nputils.py#L38) would fix this.

Alternatively, this might be clearer -- and maybe slightly faster -- if this was done with a single call to `lexsort` per dimension. Something like:
```
vars_by_dim = collections.defaultdict(list)
for d in vs:
    ...
    vars_by_dim[key].append(d)

indices = {}
for key, vars in vars_by_dim.items():
    order = np.lexsort(tuple(reversed(vars))
    indices[key] = order if ascending else order[::-1]
```